File: java/pl/edu/pg/mygame/organisms/animals/Animal.java
package pl.edu.pg.mygame.organisms.animals;

import pl.edu.pg.mygame.Grid;
import pl.edu.pg.mygame.World;
import pl.edu.pg.mygame.organisms.Organism;
import pl.edu.pg.mygame.utils.OrganismFactory;

import java.awt.Point;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static pl.edu.pg.mygame.utils.HelperFunctions.getRandomPoint;

public abstract class Animal extends Organism {
    Animal(Point p) {
        super(p);
    }


    @Override
    public void makeAction(Grid grid) {
        Point new_point = getNewPoint(grid); // this method should be overridden by some animals
        if (new_point == null)
            return;

        World.message_panel.showMessage(species + " tried to move to: " +
                String.format("(%d, %d)", new_point.x, new_point.y));
        if (grid.getOrganism(new_point) == null) {
            World.message_panel.showMessage(" and moved\n");
            grid.setOrganism(position, null);
            position = new_point;
            grid.setOrganism(position, this);
        }
        else if (grid.getOrganism(new_point).getSpecies() == this.species) {
            World.message_panel.showMessage(" and reproduced at: ");
            reproduce(new_point, grid);
        }
        else {
            World.message_panel.showMessage(" and collided with: " + grid.getOrganism(new_point).getSpecies() + " ---> ");
            handleCollision(new_point, grid);
        }
    }

    protected Point getNewPoint(Grid grid) {
        List<Point> neighbours = grid.getNeighbourCells(this.position);
        return getRandomPoint(neighbours);
    }

    private void reproduce(Point new_point, Grid grid) {
        Point dest = getRandomFreeAdjacentPoint(position, new_point, grid);
        if (dest == null) {
            World.message_panel.showMessage("null\n");
            return;
        }

        World.message_panel.showMessage(String.format("(%d, %d)", dest.x, dest.y) + "\n");

        Organism organism = OrganismFactory.create(this.species, dest);
        grid.addOrganism(organism);
    }

    public void handleCollision(Point other_org_point, Grid grid) {
        boolean collided;
        Organism other_org = grid.getOrganism(other_org_point);
        // make the collision of current organism with the other
        collided = collision(other_org, true, grid);
        // if the other organism survived, make his collision
        if (!collided)
            collided = other_org.collision(this, false, grid);


        // if nothing important happened, make the default collision
        if (!collided) {
            if (this.getStrength() >= other_org.getStrength()) {
                World.message_panel.showMessage(this.species + " killed " + other_org.getSpecies() + "\n");
                grid.setOrganism(position, null);
                position = other_org_point;
                grid.setOrganism(position, this);
            }
            else {
                World.message_panel.showMessage(other_org.getSpecies() + " killed " + this.species + "\n");
                grid.setOrganism(position, null);
            }
        }
    }


    private Point getRandomFreeAdjacentPoint(Point point1, Point point2, Grid grid) {
        Set<Point> neighbours1 = new HashSet<>(grid.getFreeNeighbourCells(point1));
        Set<Point> neighbours2 = new HashSet<>(grid.getFreeNeighbourCells(point2));
        Set<Point> neighbours_sum = new HashSet<>(neighbours1);
        neighbours_sum.addAll(neighbours2);

        if (neighbours_sum.size() == 0)
            return null;


        return getRandomPoint(new ArrayList<>(neighbours_sum));
    }
}

File: python/organisms/animals/animal.py
import copy
from abc import abstractmethod

import world
from gui.logger import Logger

from organisms.organism import Organism
from utils.helpers import Helpers


class Animal(Organism):
    @abstractmethod
    def __init__(self, pos):
        super().__init__(pos)

    def make_action(self, grid):
        new_point = copy.deepcopy(self.get_new_point(grid))
        if new_point is None:
            return

        Logger.log(self.species.name + f' tried to move to: {new_point}')
        if grid.get_organism(new_point) is None:
            Logger.log_add_to_prev(' and moved')
            grid.set_organism(self.position, None)
            self.position = new_point
            grid.set_organism(self.position, self)
        elif grid.get_organism(new_point).species == self.species:
            Logger.log_add_to_prev(' and reproduced at: ')
            self.reproduce(new_point, grid)
        else:
            Logger.log_add_to_prev(' and collided with: ' + grid.get_organism(new_point).species.name + ' ---> ')
            self.handle_collision(new_point, grid)

    def get_new_point(self, grid):
        neighbors = grid.get_neighbor_cells(self.position)
        return Helpers.get_random_point(neighbors)

    def reproduce(self, new_point, grid):
        dest = self.get_random_free_adjacent_point(self.position, new_point, grid)
        if dest is None:
            Logger.log_add_to_prev('None')
            return

        Logger.log_add_to_prev(str(dest))

        world.World.create_organism(self.species, dest, grid)

    def handle_collision(self, other_org_point, grid):
        other_org = grid.get_organism(other_org_point)
        # make the collision of current organism with the other
        collided = self.collision(other_org, True, grid)
        # if the other organism survived, make his collision
        if not collided:
            collided = other_org.collision(self, False, grid)

        # if nothing important happened, make the default collision
        if not collided:
            if self.strength >= other_org.strength:
                Logger.log_add_to_prev(self.species.name + ' killed ' + other_org.species.name)
                grid.set_organism(self.position, None)
                self.position = other_org_point
                grid.set_organism(self.position, self)
            else:
                Logger.log_add_to_prev(other_org.species.name + ' killed ' + self.species.name)
                grid.set_organism(self.position, None)

    @staticmethod
    def get_random_free_adjacent_point(point1, point2, grid):
        neighbors1 = set(grid.get_free_neighbor_cells(point1))
        neighbors2 = set(grid.get_free_neighbor_cells(point2))
        neighbors_union = neighbors1.union(neighbors2)

        if len(neighbors_union) == 0:
            return None

        return Helpers.get_random_point(list(neighbors_union))
